### PR TITLE
[PD][HeteroArch]Fix accuracy issue with CPU_ATTN as Decoder and Flash_ATTN as prefiller

### PR DESCRIPTION
**Pull request by xuechendi**
## Purpose

Fix the error reported in #38710 

When using CPU as decoder to work with Hetero Arch Platform (CUDA, Intel GPU, Intel Gaudi ...) as prefiller, there is an accuracy issue due to CPU KV layout need addition `packing` step

In this PR, we fixed this issue by following cpu_attn kv Packing method to pack FlashAttn plain KV to CPU packed KV.
The work is done in platforms/cpu.py - `pack_kv_cache()`

## Changes proposed

1. In NixlConnector, introduced a new method - `post_process_device_kv_on_receive_heterogeneous_attn`, which will be called if `self.enable_heterogeneous_attn_post_process = True`
2. `self.enable_heterogeneous_attn_post_process = True` will only be enabled when `attention_back == "CPU_ATTN"` and `remote_agent.attention_backend != "CPU_ATTN"`
3. in cpu.py, introduce a new method - "pack_kv_cache" which will be called in `post_process_device_kv_on_receive_heterogeneous_attn` to pack KV

## Test Plan
Prefill runs on CUDA / XPU
```
export VLLM_LOGGING_LEVEL=debug
export UCX_TLS=tcp
export model_name=Qwen/Qwen3-0.6B
export tp_size=1

export ZE_AFFINITY_MASK=0
# PREFILL
VLLM_NIXL_SIDE_CHANNEL_HOST=0.0.0.0 \
VLLM_NIXL_SIDE_CHANNEL_PORT=5577 \
VLLM_WORKER_MULTIPROC_METHOD=spawn \
VLLM_ENABLE_V1_MULTIPROCESSING=1 \
vllm serve $model_name \
  -tp $tp_size \
  --host 0.0.0.0 \
  --port 8100 \
  --seed 42 \
  --enforce-eager \
  --dtype float16 \
  --gpu-memory-utilization 0.2 \
  --kv-transfer-config '{"kv_connector":"NixlConnector","kv_role":"kv_both","kv_buffer_device":"cpu","kv_connector_extra_config":{"enforce_handshake_compat": false}}' \
  --max-model-len 9216 \
  --max_num_batched_tokens 2048 \
  --block-size 64 \
  --no-enable-prefix-caching \
  --disable-hybrid-kv-cache-manager 2>&1 | tee prefix_test.log &

# PROXY
python3 /workspace/vllm/tests/v1/kv_connector/nixl_integration/toy_proxy_server.py \
  --prefiller-host localhost \
  --prefiller-port 8100 \
  --decoder-host localhost \
  --decoder-port 8200 \
  --host localhost \
  --port 8300
```

Decode runs on CPU
```
export UCX_TLS=tcp
export model_name=Qwen/Qwen3-0.6B
export tp_size=1

# DECODE
VLLM_NIXL_SIDE_CHANNEL_HOST=0.0.0.0 \
VLLM_NIXL_SIDE_CHANNEL_PORT=5587 \
VLLM_WORKER_MULTIPROC_METHOD=spawn \
VLLM_ENABLE_V1_MULTIPROCESSING=1 \
vllm serve $model_name \
  -tp $tp_size \
  --host 0.0.0.0 \
  --port 8200 \
  --seed 42 \
  --enforce-eager \
  --dtype float16 \
  --gpu-memory-utilization 0.8 \
  --kv-transfer-config '{"kv_connector":"NixlConnector","kv_role":"kv_both","kv_buffer_device":"cpu","kv_connector_extra_config":{"enforce_handshake_compat": false}}' \
  --max-model-len 9216 \
  --max_num_batched_tokens 2048 \
  --block-size 64 \
  --no-enable-prefix-caching \
  --disable-hybrid-kv-cache-manager  2>&1 | tee decode_test.log
```

## Test Result
```
curl -X POST "http://localhost:8300/v1/chat/completions"   -H "Content-Type: application/json"   -d '{
    "model": "Qwen/Qwen3-0.6B",
    "messages": [
        {
        "role": "user",
        "content": "Hello, how are you?"
        }
    ],
    "max_tokens": 30, "stream": false
 }'
```

{"id":"chatcmpl-d72d90c9-5f06-4a9e-9f7b-d08c63a77fef","object":"chat.completion","created":1775238578,"model":"Qwen/Qwen3-0.6B","choices":[{"index":0,"message":{"role":"assistant","content":"**<think>\nOkay, the user just asked, \"Hello, how are you?\" I need to respond appropriately. First, I should acknowledge their greeting.**","refusal":null,"annotations":null,"audio":null,"function_call":null,"tool_calls":[],"reasoning":null},"logprobs":null,"finish_reason":"length","stop_reason":null,"token_ids":null}],"service_tier":null,"system_fingerprint":null,"usage":{"prompt_tokens":14,"total_tokens":44,"completion_tokens":30,"prompt_tokens_details":null},"prompt_logprobs":null,"prompt_token_ids":null,"kv_transfer_params":n

accuracy test:
```
lm_eval run \
  --model local-completions \
  --model_args model=Qwen/Qwen3-0.6B,base_url=http://localhost:8300/v1/completions,num_concurrent=32,tokenized_requests=False \
  --tasks gsm8k --batch_size 1 --limit 256
```

image

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

### CODE REVIEW

**Review by bigPYJ1151 — APPROVED** [review]
LGTM :)

File: vllm/distributed/kv_transfer/kv_connector/v1/nixl_connector.py [hunk 1]
```diff
@@ -2369,6 +2372,21 @@ def _validate_remote_agent_handshake(
...
+            nixl_agent_meta.attn_backend_name != self.backend_name
+            and self.backend_name in ["CPU_ATTN"]
+        ):
+            if self._is_hma_required:
+                raise RuntimeError(
+                    "heterogeneous attn post process is not supported with HMA"
+                )
+            logger.info(
+                "[Experimental] CPU_ATTN backend is used, "
+                "hint heterogeneous attn post process"
```
Comment (NickLucche) [inline vllm/distributed/kv_transfer/kv_connector/v1/nixl_connector.py:2387]: not sure this log is super clear >.<

File: vllm/distributed/kv_transfer/kv_connector/v1/nixl_connector.py [hunk 2]
```diff
@@ -2542,6 +2560,28 @@ def post_process_device_kv_on_receive(
...
+        self, block_ids: list[int]
+    ):
+        """
+        Post process device kv cache after receiving from remote
+        for heterogeneous attention.
+        """
+        assert self.enable_heterogeneous_attn_post_process
+
+        indices = torch.tensor(block_ids, device=self.device_type, dtype=torch.long)
+
```
Comment (NickLucche) [inline vllm/distributed/kv_transfer/kv_connector/v1/nixl_connector.py:2573]: I think we want to unify post-processing here soon or we'll end up with a lot of methods in the worker that don't strictly belong here @xuechendi

File: vllm/platforms/cpu.py [hunk 3]
```diff
@@ -520,3 +520,43 @@ def import_kernels(cls) -> None:
                 import vllm._C  # noqa: F401
             except ImportError as e:
                 logger.warning("Failed to import from vllm._C: %r", e)
+
+    @classmethod
+    def pack_kv_cache(
+        cls,
```
Comment (NickLucche) [inline vllm/platforms/cpu.py:526]: I believe this method name should mention the fact that it's an in-place update of the kv cache because that's pretty important.
